2-WebApp-graph-user/2-5-HybridFlow/Views/Shared/_Layout.cshtml

```diff
@@ -0,0 +1,217 @@
+                        <li class="nav-item">
+                            <a class="nav-link text-dark" asp-area="" asp-controller="Home" asp-action="Index">Home</a>
+                        </li>
+
+                        @if(User.Identity is not null && User.Identity.IsAuthenticated) {
+                            <li class="nav-item">
+                                <a class="nav-link text-dark" asp-area="" asp-controller="Home" asp-action="Profile">Profile</a>
+                            </li>
+                        }
+
+                        <li class="nav-item">
+                            <a class="nav-link text-dark" asp-area="" asp-controller="Home" asp-action="Privacy">Privacy</a>
+                        </li>
+                    </ul>
+                </div>
+            </div>
+        </nav>
+    </header>
+    <div class="container">
+        <main role="main" class="pb-3">
+            @RenderBody()
+        </main>
+    </div>
+
+    <footer class="border-top footer text-muted">
+        <div class="container">
+            &copy; 2022 - WebApp_OpenIDConnect_DotNet - <a asp-area="" asp-controller="Home" asp-action="Privacy">Privacy</a>
+        </div>
+    </footer>
+    <script src="~/lib/jquery/dist/jquery.min.js"></script>
+    <script src="~/lib/bootstrap/dist/js/bootstrap.bundle.min.js"></script>
+    <script>
+        (function () {
+            const scopes = 
+            @{
+                var apiScopes = Configuration["DownstreamApi:Scopes"].Split(' ');
+                @Html.Raw("[");
+
+                foreach(var scope in apiScopes) {
+                    @Html.Raw($"'{scope}',")
+                }
+
+                @Html.Raw("];");
+
+                Context.Session.TryGetValue("Microsoft.Identity.Hybrid.Authentication", out var spaCode);
+
+                if (spaCode is not null && !string.IsNullOrEmpty(Encoding.Default.GetString(spaCode)))
+                {
+                    @Html.Raw($"const code = '{Encoding.Default.GetString(spaCode)}';");
+                    Context.Session.Remove("Microsoft.Identity.Hybrid.Authentication");
+                }
+                else
+                {
+                    @Html.Raw($"const code = '';");
+                }
+            }
+
+            if (!!code) {
+                msalInstance
+                    .handleRedirectPromise()
+                    .then(result => {
+                        if (result) {
+                            console.log('MSAL: Returning from login');
+                            document.dispatchEvent(new Event('AUTHENTICATED'));
+                            return result;
+                        }
+
+                        const timeLabel = "Time for acquireTokenByCode";
+                        console.time(timeLabel);
+                        console.log('MSAL: acquireTokenByCode hybrid parameters present');
+
+                        return msalInstance.acquireTokenByCode({
+                            code,
```

Comment:
One other consideration (which you can do incrementally) is to not redeem the same `code` twice (e.g. by refreshing the page).